### PR TITLE
Support Exporting Bitshift to ONNX

### PR DESCRIPTION
**Pull request by lara-hdr**
Support exporting left/right bitshifts to ONNX for all opset versions.

ONNX has a bitshift operator in opset 11, but it only supports unsigned ints, so it can't be used in PyTorch (since only uint8 is the only uint type).

### CODE REVIEW

**Review by spandantiwari — CHANGES_REQUESTED** [review]
Could you please confirm if onnx::bitshift can be used here or not.

**Comment by neginraoof** [comment]
CI failure is TestONNXRuntime_opset11::test_bitshift_uint8 FAILED

**Comment by lara-hdr** [comment]
thank you @neginraoof, I had forgotten about this one.  I updated the ORT version for the test.

File: torch/onnx/symbolic_opset11.py [hunk 1]
```diff
@@ -356,3 +356,25 @@ def index_copy(g, self, dim, index, source):
         return g.op("ATen", self, index, source, dim_i=dim_value, operator_s="index_copy")
     expanded_index_shape, expanded_index = sym_help._index_fill_reshape_helper(g, self, dim, index)
     return scatter(g, self, dim, expanded_index, source)
+
+
+def __rshift_(g, self, other):
+    if self.type().scalarType() == 'Byte':
+        if other.type().scalarType != 'Byte':
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset11.py]: nit: one scalarType has parentheses the other does not ...

File: torch/onnx/symbolic_opset11.py [hunk 2]
```diff
@@ -356,3 +356,25 @@ def index_copy(g, self, dim, index, source):
...
+            other = g.op("Cast", other, to_i=sym_help.cast_pytorch_to_onnx['Byte'])
+        return g.op('BitShift', self, other, direction_s="RIGHT")
+    two = g.op('Constant', value_t=torch.tensor(2, dtype=torch.float))
+    two_pow = g.op('Pow', two, other)
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset11.py]: There is probably an edge case not coverred. When `self` is integral, and `other` is floating point, pytorch will cast `other` to integral as well. 

Please try
```python
        class BitshiftModel(torch.nn.Module):
            def forward(self, input):
                return input << 2.4
        input = torch.arange(24, dtype=torch.long).reshape(3, 4, 2)
        self.run_test(BitshiftModel(), input)
```

**Review by BowenBao — CHANGES_REQUESTED** [review]

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 3]
```diff
@@ -655,6 +655,25 @@ def forward(self, input):
         model = StandardDeviation()
         self.run_test(model, x)
 
+    def test_bitshift(self):
```
[the remark this answers is not in the document]
Comment (lara-hdr) [reply test/onnx/test_pytorch_onnx_onnxruntime.py]: I updated the tests to take it into account

File: torch/onnx/symbolic_opset11.py (hunk 2 above)
Comment (lara-hdr) [reply torch/onnx/symbolic_opset11.py]: good catch. I updated the code to take this into account.

**Comment by neginraoof** [comment]
Thanks. Looks good.

File: torch/onnx/symbolic_opset9.py [hunk 4]
```diff
@@ -810,6 +810,20 @@ def __or_(g, input, other):
     return g.op('Or', input, other)
 
 
+def __rshift_(g, self, other):
```
[the remark this answers is not in the document]
Comment (spandantiwari) [reply torch/onnx/symbolic_opset9.py]: @lara-hdr - is this resolved? Can we not use onnx:;bitshift here?

File: torch/onnx/symbolic_opset9.py [hunk 5]
```diff
@@ -827,6 +827,38 @@ def __or_(g, input, other):
...
+        other = g.op("Cast", other, to_i=sym_help.cast_pytorch_to_onnx['Float'])
+    two_pow = g.op('Pow', two, other)
+
+    rshift = g.op('Div', self, two_pow)
+    return rshift
+
+
+def __lshift_(g, self, other):
+    # make sure to cast other to self's type
+    # (when slef is long, make sure that other is not float)
```
Comment (BowenBao) [inline torch/onnx/symbolic_opset9.py]: nit: typo .. 
```suggestion
    # (when self is long, make sure that other is not float)
```

**Review by BowenBao — APPROVED** [review]
LGTM, thanks.

**Review by spandantiwari — APPROVED** [review]
LGTM. Please address a single comment that I have.

**Comment by lara-hdr** [comment]
@spandantiwari, yes bitshift is used here https://github.com/pytorch/pytorch/pull/28210/files?file-filters%5B%5D=.py#diff-e5e8501daccb885bc428bcf6c54fd98eR383 (in opset 11)

**Review by houseroad — APPROVED** [review]
LGTM

**Pull request closed without merging**